src/torrent.rs

```diff
@@ -108,86 +115,133 @@ impl Torrent {
             requested_blocks: 0,
             percent_complete: 0.0,
             repeated_blocks: HashMap::new(),
+            in_progress_blocks: vec![],
+            completed_pieces: (0..number_of_pieces)
+                .map(|_pi| (0..number_of_blocks).map(|_bi| None).collect())
+                .collect(),
         }
     }
 
-    pub fn get_next_block(&mut self, bitfield: &BitField) -> Option<(u32, u32, u32)> {
-        for p in self.pieces.iter_mut() {
-            // is this piece one that is present in the bitfield?
-            if bitfield.is_set(p.index as usize).is_ok() {
-                for b in p.blocks.iter_mut() {
-                    match b.state {
-                        BlockState::Done => continue,
-                        BlockState::Requested => {
-                            if self.percent_complete > END_GAME_PROGRESS_THRESHOLD {
-                                return Some((p.index, b.offset, FIXED_BLOCK_SIZE));
-                            }
-                            let now = Instant::now();
-                            let last = b.last_request.unwrap();
-                            if now - last > REQUEST_WAIT_TIME {
-                                b.last_request = Some(now);
-                                return Some((p.index, b.offset, FIXED_BLOCK_SIZE));
-                            } else {
-                                continue;
-                            }
-                        }
-                        BlockState::NotRequested => {
-                            b.state = BlockState::Requested;
-                            b.last_request = Some(Instant::now());
-                            self.requested_blocks += 1;
-                            return Some((p.index, b.offset, FIXED_BLOCK_SIZE));
-                        }
+    pub fn get_next_block(&mut self, bitfield: &BitField) -> Option<PieceIndexOffsetLength> {
+        if self.in_progress_blocks.len() == 256 {
+            // there are no more blocks for the requester to help with "right now"
+            println!(
+                "we are at capacity for new in progress blocks; current in progress: {:?}",
+                self.in_progress_blocks
+                    .iter()
+                    .map(|block| { (block.piece_index, block.offset) })
+            );
+            return None;
+        }
+
+        let res: Option<(u32, &mut VecDeque<Block>)> = {
+            let mut res = None;
+            // O(total number of pieces); always pulls pieces and blocks based on exact order of index of piece from 0 to total number of pieces
```

Comment:
I need to be able to design, implement, and swap out multiple ways of deciding which block to grab.